moveos/moveos/src/moveos.rs

```diff
@@ -267,14 +309,104 @@ impl MoveOS {
         match result {
             Ok(return_values) => {
                 // if execute success, finish the session to check if it change the state
-                match session.finish_with_extensions(VMStatus::Executed) {
+                match session.finish_with_extensions(KeptVMStatus::Executed) {
                     Ok(_) => FunctionResult::ok(return_values),
                     Err(e) => FunctionResult::err(e),
                 }
             }
             Err(e) => FunctionResult::err(e),
         }
     }
+
+    // Execute use action with pre_execute and post_execute.
+    // Return the user action execution status if success,
+    // else return VMError and a bool which indicate if we should respawn the session.
+    fn execute_user_action(
+        &self,
+        session: &mut MoveOSSession<'_, '_, MoveOSResolverProxy<MoveOSStore>, MoveOSGasMeter>,
+        action: VerifiedMoveAction,
+        pre_execute_functions: Vec<FunctionCall>,
+        post_execute_functions: Vec<FunctionCall>,
+    ) -> Result<VMStatus, (VMError, bool)> {
+        // user pre_execute
+        // If the pre_execute failed, we finish the session directly and return the TransactionOutput.
+        session
+            .execute_function_call(pre_execute_functions, true)
+            .map_err(|e| (e, false))?;
+
+        // execute main tx
+        let execute_result = session.execute_move_action(action);
+        let vm_status = vm_status_of_result(execute_result);
+
+        // If the user action or post_execute failed, we need respawn the session,
+        // and execute system_pre_execute, system_post_execute and user pre_execute, user post_execute.
+        let status = match vm_status.clone().keep_or_discard() {
+            Ok(status) => {
+                if status != KeptVMStatus::Executed {
+                    return Err((
+                        // The VMError will not be used.
+                        PartialVMError::new(StatusCode::UNKNOWN_STATUS).finish(Location::Undefined),
+                        true,
+                    ));
+                }
+                session
+                    .execute_function_call(post_execute_functions, true)
+                    .map_err(|e| (e, true))?;
```

Comment:
If the post_execute_functions failed and `need_respawn = true` here, respawn a new session to re-execute the post_execute_functions, it is likely to fail again. So `need_respawn` should be false here?